man/sd-daemon.3
'\" t
.TH "SD\-DAEMON" "3" "" "systemd 7" "sd-daemon"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
sd-daemon, SD_EMERG, SD_ALERT, SD_CRIT, SD_ERR, SD_WARNING, SD_NOTICE, SD_INFO, SD_DEBUG \- Reference implementation of APIs for new\-style daemons
.SH "SYNOPSIS"
.sp
.ft B
.nf
#include <systemd/sd\-daemon\&.h>
.fi
.ft
.HP \w'\fBpkg\-config\ \-\-cflags\ \-\-libs\ libsystemd\-daemon\fR\ 'u
\fBpkg\-config \-\-cflags \-\-libs libsystemd\-daemon\fR
.SH "DESCRIPTION"
.PP
sd\-daemon\&.c
and
sd\-daemon\&.h
provide a reference implementation of various APIs for new\-style daemons, as implemented by the
\fBsystemd\fR(1)
init system\&.
.PP
See
\fBsd_listen_fds\fR(3),
\fBsd_notify\fR(3),
\fBsd_booted\fR(3),
\fBsd_is_fifo\fR(3)
for more information about the functions implemented\&. In addition to these functions a couple of logging prefixes are defined as macros:
.sp
.if n \{\
.RS 4
.\}
.nf
#define SD_EMERG   "<0>"  /* system is unusable */
#define SD_ALERT   "<1>"  /* action must be taken immediately */
#define SD_CRIT    "<2>"  /* critical conditions */
#define SD_ERR     "<3>"  /* error conditions */
#define SD_WARNING "<4>"  /* warning conditions */
#define SD_NOTICE  "<5>"  /* normal but significant condition */
#define SD_INFO    "<6>"  /* informational */
#define SD_DEBUG   "<7>"  /* debug\-level messages */
.fi
.if n \{\
.RE
.\}
.PP
These prefixes are intended to be used in conjunction with STDERR\-based logging as implemented by systemd\&. If a systemd service definition file is configured with
\fIStandardError=syslog\fR
or
\fIStandardError=kmsg\fR, these prefixes can be used to encode a log level in lines printed\&. This is similar to the kernel
\fBprintk()\fR\-style logging\&. See
\fBklogctl\fR(2)
for more information\&.
.PP
The log levels are identical to
\fBsyslog\fR(3)\*(Aqs log level system\&. To use these prefixes simply prefix every line with one of these strings\&. A line that is not prefixed will be logged at the default log level SD_INFO\&.
.PP
\fBExample\ \&1.\ \&Hello World\fR
.PP
A daemon may log with the log level NOTICE by issuing this call:
.sp
.if n \{\
.RS 4
.\}
.nf
fprintf(stderr, SD_NOTICE "Hello World!\en");
.fi
.if n \{\
.RE
.\}
.SH "NOTES"
.PP
These interfaces are provided by the reference implementation of APIs for new\-style daemons and distributed with the systemd package\&. The algorithms they implement are simple, and can easily be reimplemented in daemons if it is important to support this interface without using the reference implementation\&. See the respective function man pages for details\&.
.PP
In addition, for details about the algorithms check the liberally licensed reference implementation sources:
\m[blue]\fB\%http://cgit.freedesktop.org/systemd/systemd/plain/src/libsystemd-daemon/sd-daemon.c\fR\m[]
and
\m[blue]\fB\%http://cgit.freedesktop.org/systemd/systemd/plain/src/systemd/sd-daemon.h\fR\m[]
.PP
These APIs are implemented in the reference implementation\*(Aqs
sd\-daemon\&.c
and
sd\-daemon\&.h
files\&. These interfaces are available as shared library, which can be compiled and linked to with the
\fBlibsystemd\-daemon\fR\ \&\fBpkg-config\fR(1)
file\&. Alternatively, applications consuming these APIs may copy the implementation into their source tree, either verbatim or in excerpts\&.
.PP
The functions directly related to new\-style daemons become NOPs when \-DDISABLE_SYSTEMD is set during compilation and the reference implementation is used as drop\-in files\&. In addition, if
sd\-daemon\&.c
is compiled on non\-Linux systems they become NOPs\&.
.SH "SEE ALSO"
.PP
\fBsystemd\fR(1),
\fBsd_listen_fds\fR(3),
\fBsd_notify\fR(3),
\fBsd_booted\fR(3),
\fBsd_is_fifo\fR(3),
\fBdaemon\fR(7),
\fBsystemd.service\fR(5),
\fBsystemd.socket\fR(5),
\fBfprintf\fR(3),
\fBsd-readahead\fR(3),
\fBpkg-config\fR(1)


NEWS
uselessd System and Service Manager

Version 8 (2014-11-16)
----------------------
   *  A new feature has been introduced - hard job failures for uselessd
      conditions. By default, if a Condition*= directive in a systemd
      or uselessd service is false, the job is skipped. On the other hand,
      if the file /run/systemd/condition-job-failure exists with a text
      string of Y, T, yes or true, all false conditions will lead to a
      job failure being logged for units on a global level. This is
      similar to systemd-git's (presumably will be in 218) Assert*=
      options, except implemented as a global and volatile flag in the
      /run directory. The default tmpfiles.d configuration creates the
      file with a value of N.

   *  The default tmpfiles.d configuration now respects the legacy
      /forcefsck, /fastboot and /forcequotacheck files and will not
      interfere with them existing.

   *  systemd-sleep has been replaced with a shell script called
      uselessd-sleep. Much simpler, it is currently limited - it
      only executes the /usr/lib/systemd/system-sleep directory once
      and expects shell scripts with an *.sh extension only. In addition,
      it requires an explicitly configured /etc/systemd/sleep.conf file.
      As a result, we provide a stock one. Relevant units have been
      updated to use uselessd-sleep.

   *  systemd-shutdown and systemd-shutdownd will no longer be built if
      --disable-initd is passed to the configure script.

   *  A new configure switch has been introduced: --disable-update-utmp,
      for stripping out the systemd-update-utmp tool from the build.

   *  main-no-init has been significantly simplified. In keeping with
      its purpose as a stage 2 process manager/supervisor, it no longer
      handles things irrelevant to it: /proc/cmdline parsing, crash
      shells, crash VTs, watchdog timers, switch-root handling, daemon
      reexec/keexec/shutdown/reboot/halt/poweroff. In addition, it
      no longer dumps core by default and the --user option has been
      removed, due to main-no-init's intended usage being as a global
      system instance.

      System generator directories are no longer managed and the
      SIGINT/SIGTERM signals are interpreted as a MANAGER_EXIT.

   *  Some updates to systemd-tmpfiles, such as following symlinks
      in lines with w commands and parsing prefixes before suffixes in
      e.g. directories. No major post-208 features as of yet, however.

   *  A few new test cases have been imported from upstream systemd
      with some modifications: test-async, test-set and test-ratelimit.
      More cases for test-util have been added, as well.

   *  Miscellaneous bugfixes and source refactoring.

Version 7 (2014-11-07)
----------------------
   *  A new program has been introduced - uselessd-cgls, which
      is a lightweight alternative to systemd-cgls written in
      Bourne shell. It can be configured out of the build with
      --disable-cgls and comes with a manpage.

   *  Two other new configure switches have been introduced:
      --disable-remount-fs and --disable-sleep, to, respectively
      strip out the systemd-remount-fs tool for rescanning /etc/fstab,
      and the systemd-sleep tool for controlling sleep mode from sysfs
      plus applying user configuration (as well as all respective units
      and manpages).

   *  A wide variety of bugfixes.

Version 6 (2014-10-29)
----------------------
   *  Three new configure options have been added:

      --enable-mudflap for building with GCC Mudflap compiler
      instrumentation on compilers that support it.

      --disable-plymouth for turning off Plymouth event integration
      in units and tty-ask-password-agent.

      --disable-password-agents to strip the systemd-ask-password,
      systemd-reply-password and systemd-tty-ask-password-agent tools
      out of the build. Note that the path units will still be shipped
      if HAVE_SYSV_COMPAT evaluates to true.

   *  The KillMode=mixed unit option for sending SIGTERM to the main
      process of the control group and then a SIGKILL to all the
      remaining ones in the same group, has been backported. The exact
      kill signals are configurable, see man 5 systemd.kill for details.

   *  systemctl has gained an --avoid-bus option which will head
      straight for the /run/systemd/private IPC socket and not do
      any system bus checks, where such a condition is allowed.
      This is a small preliminary step for an intended future
      overhaul of the IPC system.

   *  Some dead code has been removed, shutdownd will no longer
      run an unmount in containers (instead relying on the kernel's
      namespace destructors), main-no-init no longer checks for
      Plymouth running.

Version 5 (2014-10-22)
----------------------
   *  Preliminary experimental support for running a system-wide
      instance of uselessd without being PID1 (init) has been
      introduced.
      See http://uselessd.darknedgy.net/PidNone for details.

   *  Don't allow enabling of masked units.

   *  More vigorous overflow checking in strv_push().

Version 4 (2014-10-06)
----------------------
   *  systemd-analyze, -activate, -cgtop and -delta can now be
      configured out of the build.

   *  systemd-initctl is now only installed when the HAVE_SYSV_COMPAT
      conditional evaluates to true.

   *  The infamous /proc/cmdline "debug" argument fiasco has been
      addressed by renaming it to "systemd.debug", so that it uses
      an exclusive namespace. Currently, upstream systemd resolves
      the issue by keeping the generic term "debug", but simply
      redirecting output to journald as soon as it is available.
      As we do not have it, we implement this.

   *  Various bugfixes from systemd-stable have been backported, like
      basic.target and sysinit.target no longer having a
      RefuseManualStart option.

Version 3 (2014-09-30)
----------------------

src/libsystemd-daemon/sd-daemon.c
/*-*- Mode: C; c-basic-offset: 8; indent-tabs-mode: nil -*-*/

/***
  Copyright 2010 Lennart Poettering

  Permission is hereby granted, free of charge, to any person
  obtaining a copy of this software and associated documentation files
  (the "Software"), to deal in the Software without restriction,
  including without limitation the rights to use, copy, modify, merge,
  publish, distribute, sublicense, and/or sell copies of the Software,
  and to permit persons to whom the Software is furnished to do so,
  subject to the following conditions:

  The above copyright notice and this permission notice shall be
  included in all copies or substantial portions of the Software.

  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
  EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
  MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
  NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
  BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
  ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
  CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
  SOFTWARE.
***/

#ifndef _GNU_SOURCE
#  define _GNU_SOURCE
#endif

#include <sys/types.h>
#include <sys/stat.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <fcntl.h>
#include <netinet/in.h>
#include <stdlib.h>
#include <errno.h>
#include <unistd.h>
#include <string.h>
#include <stdarg.h>
#include <stdio.h>
#include <stddef.h>
#include <limits.h>

#if defined(__linux__) && !defined(SD_DAEMON_DISABLE_MQ)
#  include <mqueue.h>
#endif

#include "sd-daemon.h"

#if (__GNUC__ >= 4)
#  ifdef SD_EXPORT_SYMBOLS
/* Export symbols */
#    define _sd_export_ __attribute__ ((visibility("default")))
#  else
/* Don't export the symbols */
#    define _sd_export_ __attribute__ ((visibility("hidden")))
#  endif
#else
#  define _sd_export_
#endif

_sd_export_ int sd_listen_fds(int unset_environment) {

#if defined(DISABLE_SYSTEMD) || !defined(__linux__)
        return 0;
#else
        int r, fd;
        const char *e;
        char *p = NULL;
        unsigned long l;

        e = getenv("LISTEN_PID");
        if (!e) {
                r = 0;
                goto finish;
        }

        errno = 0;
        l = strtoul(e, &p, 10);

        if (errno > 0) {
                r = -errno;
                goto finish;
        }

        if (!p || p == e || *p || l <= 0) {
                r = -EINVAL;
                goto finish;
        }

        /* Is this for us? */
        if (getpid() != (pid_t) l) {
                r = 0;
                goto finish;
        }

        e = getenv("LISTEN_FDS");
        if (!e) {
                r = 0;
                goto finish;
        }

        errno = 0;
        l = strtoul(e, &p, 10);

        if (errno > 0) {
                r = -errno;
                goto finish;
        }

        if (!p || p == e || *p) {
                r = -EINVAL;
                goto finish;
        }

        for (fd = SD_LISTEN_FDS_START; fd < SD_LISTEN_FDS_START + (int) l; fd ++) {
                int flags;

                flags = fcntl(fd, F_GETFD);
                if (flags < 0) {
                        r = -errno;
                        goto finish;
                }

                if (flags & FD_CLOEXEC)
                        continue;

                if (fcntl(fd, F_SETFD, flags | FD_CLOEXEC) < 0) {
                        r = -errno;
                        goto finish;
                }
        }

        r = (int) l;

finish:
        if (unset_environment) {
                unsetenv("LISTEN_PID");
                unsetenv("LISTEN_FDS");
        }

        return r;
#endif
}

_sd_export_ int sd_is_fifo(int fd, const char *path) {
        struct stat st_fd;

        if (fd < 0)
                return -EINVAL;

        if (fstat(fd, &st_fd) < 0)
                return -errno;

        if (!S_ISFIFO(st_fd.st_mode))
                return 0;

        if (path) {
                struct stat st_path;

                if (stat(path, &st_path) < 0) {

                        if (errno == ENOENT || errno == ENOTDIR)
                                return 0;

                        return -errno;
                }

                return
                        st_path.st_dev == st_fd.st_dev &&
                        st_path.st_ino == st_fd.st_ino;
        }

        return 1;
}

_sd_export_ int sd_is_special(int fd, const char *path) {
        struct stat st_fd;

        if (fd < 0)
                return -EINVAL;

        if (fstat(fd, &st_fd) < 0)
                return -errno;

        if (!S_ISREG(st_fd.st_mode) && !S_ISCHR(st_fd.st_mode))
                return 0;

        if (path) {
                struct stat st_path;

                if (stat(path, &st_path) < 0) {

                        if (errno == ENOENT || errno == ENOTDIR)
                                return 0;

                        return -errno;
                }

                if (S_ISREG(st_fd.st_mode) && S_ISREG(st_path.st_mode))
                        return
                                st_path.st_dev == st_fd.st_dev &&
                                st_path.st_ino == st_fd.st_ino;
                else if (S_ISCHR(st_fd.st_mode) && S_ISCHR(st_path.st_mode))
                        return st_path.st_rdev == st_fd.st_rdev;
                else
                        return 0;
        }

        return 1;
}

static int sd_is_socket_internal(int fd, int type, int listening) {
        struct stat st_fd;

        if (fd < 0 || type < 0)
                return -EINVAL;

        if (fstat(fd, &st_fd) < 0)
                return -errno;

        if (!S_ISSOCK(st_fd.st_mode))
                return 0;

        if (type != 0) {
                int other_type = 0;
                socklen_t l = sizeof(other_type);

                if (getsockopt(fd, SOL_SOCKET, SO_TYPE, &other_type, &l) < 0)
                        return -errno;

                if (l != sizeof(other_type))
                        return -EINVAL;

                if (other_type != type)
                        return 0;
        }

        if (listening >= 0) {
                int accepting = 0;
                socklen_t l = sizeof(accepting);

                if (getsockopt(fd, SOL_SOCKET, SO_ACCEPTCONN, &accepting, &l) < 0)
                        return -errno;

                if (l != sizeof(accepting))
                        return -EINVAL;

                if (!accepting != !listening)
                        return 0;
        }

        return 1;
}

union sockaddr_union {
        struct sockaddr sa;
        struct sockaddr_in in4;
        struct sockaddr_in6 in6;
        struct sockaddr_un un;
        struct sockaddr_storage storage;
};

_sd_export_ int sd_is_socket(int fd, int family, int type, int listening) {
        int r;

        if (family < 0)
                return -EINVAL;

        r = sd_is_socket_internal(fd, type, listening);
        if (r <= 0)
                return r;

        if (family > 0) {
                union sockaddr_union sockaddr = {};
                socklen_t l = sizeof(sockaddr);

                if (getsockname(fd, &sockaddr.sa, &l) < 0)
                        return -errno;

                if (l < sizeof(sa_family_t))
                        return -EINVAL;

                return sockaddr.sa.sa_family == family;
        }

        return 1;
}

_sd_export_ int sd_is_socket_inet(int fd, int family, int type, int listening, uint16_t port) {
        union sockaddr_union sockaddr = {};
        socklen_t l = sizeof(sockaddr);
        int r;

        if (family != 0 && family != AF_INET && family != AF_INET6)
                return -EINVAL;

        r = sd_is_socket_internal(fd, type, listening);
        if (r <= 0)
                return r;

        if (getsockname(fd, &sockaddr.sa, &l) < 0)
                return -errno;

        if (l < sizeof(sa_family_t))
                return -EINVAL;

        if (sockaddr.sa.sa_family != AF_INET &&
            sockaddr.sa.sa_family != AF_INET6)
                return 0;

        if (family > 0)
                if (sockaddr.sa.sa_family != family)
                        return 0;

        if (port > 0) {
                if (sockaddr.sa.sa_family == AF_INET) {
                        if (l < sizeof(struct sockaddr_in))
                                return -EINVAL;

                        return htons(port) == sockaddr.in4.sin_port;
                } else {
                        if (l < sizeof(struct sockaddr_in6))
                                return -EINVAL;

                        return htons(port) == sockaddr.in6.sin6_port;
                }
        }

        return 1;
}

_sd_export_ int sd_is_socket_unix(int fd, int type, int listening, const char *path, size_t length) {
        union sockaddr_union sockaddr = {};
        socklen_t l = sizeof(sockaddr);
        int r;

        r = sd_is_socket_internal(fd, type, listening);
        if (r <= 0)
                return r;

        if (getsockname(fd, &sockaddr.sa, &l) < 0)
                return -errno;

        if (l < sizeof(sa_family_t))
                return -EINVAL;

        if (sockaddr.sa.sa_family != AF_UNIX)
                return 0;

        if (path) {
                if (length == 0)
                        length = strlen(path);

                if (length == 0)
                        /* Unnamed socket */
                        return l == offsetof(struct sockaddr_un, sun_path);

                if (path[0])
                        /* Normal path socket */
                        return
                                (l >= offsetof(struct sockaddr_un, sun_path) + length + 1) &&
                                memcmp(path, sockaddr.un.sun_path, length+1) == 0;
                else
                        /* Abstract namespace socket */
                        return
                                (l == offsetof(struct sockaddr_un, sun_path) + length) &&
                                memcmp(path, sockaddr.un.sun_path, length) == 0;
        }

        return 1;
}

_sd_export_ int sd_is_mq(int fd, const char *path) {
#if !defined(__linux__) || defined(SD_DAEMON_DISABLE_MQ)
        return 0;
#else
        struct mq_attr attr;

        if (fd < 0)
                return -EINVAL;

        if (mq_getattr(fd, &attr) < 0)
                return -errno;

        if (path) {
                char fpath[PATH_MAX];
                struct stat a, b;

                if (path[0] != '/')
                        return -EINVAL;

                if (fstat(fd, &a) < 0)
                        return -errno;

                strncpy(stpcpy(fpath, "/dev/mqueue"), path, sizeof(fpath) - 12);
                fpath[sizeof(fpath)-1] = 0;

                if (stat(fpath, &b) < 0)
                        return -errno;

                if (a.st_dev != b.st_dev ||
                    a.st_ino != b.st_ino)
                        return 0;
        }

        return 1;
#endif
}

_sd_export_ int sd_pid_notify(pid_t pid, int unset_environment, const char *state) {
#if defined(DISABLE_SYSTEMD) || !defined(__linux__) || !defined(SOCK_CLOEXEC)
        return 0;
#else
        union sockaddr_union sockaddr = {};
        int fd = -1;
        struct msghdr msghdr = {};
        struct iovec iovec = {};
        const char *e;
        union {
                struct cmsghdr cmsghdr;
                uint8_t buf[CMSG_SPACE(sizeof(struct ucred))];
        } control = {};
        int r;

        if (!state) {
                r = -EINVAL;
                goto finish;
        }

        e = getenv("NOTIFY_SOCKET");
        if (!e)
                return 0;

        /* Must be an abstract socket, or an absolute path */
        if ((e[0] != '@' && e[0] != '/') || e[1] == 0) {
                r = -EINVAL;
                goto finish;
        }

        fd = socket(AF_UNIX, SOCK_DGRAM|SOCK_CLOEXEC, 0);
        if (fd < 0) {
                r = -errno;
                goto finish;
        }

        sockaddr.sa.sa_family = AF_UNIX;
        strncpy(sockaddr.un.sun_path, e, sizeof(sockaddr.un.sun_path));

        if (sockaddr.un.sun_path[0] == '@')
                sockaddr.un.sun_path[0] = 0;

        iovec.iov_base = (char*) state;
        iovec.iov_len = strlen(state);

        msghdr.msg_name = &sockaddr;
        msghdr.msg_namelen = offsetof(struct sockaddr_un, sun_path) + strlen(e);

        if (msghdr.msg_namelen > sizeof(struct sockaddr_un))
                msghdr.msg_namelen = sizeof(struct sockaddr_un);

        msghdr.msg_iov = &iovec;
        msghdr.msg_iovlen = 1;

        if (pid != 0 && pid != getpid()) {
                struct cmsghdr *cmsg;
                struct ucred ucred = {};

                msghdr.msg_control = &control;
                msghdr.msg_controllen = sizeof(control);

                cmsg = CMSG_FIRSTHDR(&msghdr);
                cmsg->cmsg_level = SOL_SOCKET;
                cmsg->cmsg_type = SCM_CREDENTIALS;
                cmsg->cmsg_len = CMSG_LEN(sizeof(struct ucred));

                ucred.pid = pid;
                ucred.uid = getuid();
                ucred.gid = getgid();

                memcpy(CMSG_DATA(cmsg), &ucred, sizeof(struct ucred));
                msghdr.msg_controllen = cmsg->cmsg_len;
        }

        /* First try with fake ucred data, as requested */
        if (sendmsg(fd, &msghdr, MSG_NOSIGNAL) >= 0) {
                r = 1;
                goto finish;
        }

        /* If that failed, try with our own instead */
        if (msghdr.msg_control) {
                msghdr.msg_control = NULL;
                msghdr.msg_controllen = 0;

                if (sendmsg(fd, &msghdr, MSG_NOSIGNAL) >= 0) {
                        r = 1;
                        goto finish;
                }
        }

        r = -errno;

finish:
        if (unset_environment)
                unsetenv("NOTIFY_SOCKET");

        return r;
#endif
}

_sd_export_ int sd_pid_notifyf(pid_t pid, int unset_environment, const char *format, ...) {
#if defined(DISABLE_SYSTEMD) || !defined(__linux__)
        return 0;
#else
        char *p = NULL;
        int r;

        if (format) {
                va_list ap;

                va_start(ap, format);
                r = vasprintf(&p, format, ap);
                va_end(ap);

                if (r < 0 || !p)
                        return -ENOMEM;
        }

        return sd_pid_notify(pid, unset_environment, p);
#endif
}

_sd_export_ int sd_notify(int unset_environment, const char *state) {
#if defined(DISABLE_SYSTEMD) || !defined(__linux__) || !defined(SOCK_CLOEXEC)
        return 0;
#else
        return sd_pid_notify(0, unset_environment, state);
#endif
}

_sd_export_ int sd_notifyf(int unset_environment, const char *format, ...) {
#if defined(DISABLE_SYSTEMD) || !defined(__linux__)
        return 0;
#else
        char *p = NULL;
        int r;

        if (format) {
                va_list ap;

                va_start(ap, format);
                r = vasprintf(&p, format, ap);
                va_end(ap);

                if (r < 0 || !p)
                        return -ENOMEM;
        }

        return sd_pid_notify(0, unset_environment, p);
#endif
}

_sd_export_ int sd_booted(void) {
#if defined(DISABLE_SYSTEMD) || !defined(__linux__)
        return 0;
#else
        struct stat st;

        /* We test whether the runtime unit file directory has been
         * created. This takes place in mount-setup.c, so is
         * guaranteed to happen very early during boot. */

        if (lstat("/run/systemd/system/", &st) < 0)
                return 0;

        return !!S_ISDIR(st.st_mode);
#endif
}
